critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -241,41 +244,59 @@ public void run() {
-                log.info("write error for storage adapter put: " + wre.getMessage());
-                throw wre;
-            } catch (TransientException | IOException te) {
-                // ReadException will be caught under the IOException
-                // could be prepare or put throwing this error
-                // will move to next url
-                log.info("transient exception." + te.getMessage());
-
-            } catch (ResourceNotFoundException | ResourceAlreadyExistsException  badURL) {
-                log.info("removing URL " + u + " from list: " + badURL.getMessage());
-                urlIterator.remove();
-
+        if (urlIterator.hasNext()) {
+            while (urlIterator.hasNext()) {
+                URL u = urlIterator.next();
+                log.debug("trying " + u);
+
+                try {
+                    ByteArrayOutputStream dest = new ByteArrayOutputStream();
+                    HttpGet get = new HttpGet(u, dest);
+                    get.prepare();
+
+                    // Check content checksum and length to verify this
+                    // artifact can be synced currently
+                    Artifact curArtifact = artDao.get(artifact);
+                    log.debug("content checksums: " + get.getContentMD5() + "\n" + curArtifact.getContentChecksum().getSchemeSpecificPart());
+                    if (!get.getContentMD5().equals(curArtifact.getContentChecksum().getSchemeSpecificPart())) {
+                        throw new PreconditionFailedException("mismatched content checksum. " + artifact);
+                    }
+                    if (get.getContentLength() != curArtifact.getContentLength()) {
+                        throw new PreconditionFailedException("mismatched content length. " + artifact);
+                    }
+
+                    log.debug("artifact checksum and content length match, continue to put.");
+
+                    NewArtifact a = new NewArtifact(artifact);
+                    storageMeta = sa.put(a, get.getInputStream());
+                    log.debug("storage meta returned: " + storageMeta.getStorageLocation());
+                    return storageMeta;
+
+                } catch (WriteException wre) {
+                    // IOException will capture this if not explicitly caught
+                    log.info("write error for storage adapter put: " + wre.getMessage());
+                    throw wre;
+                } catch (TransientException | IOException te) {
+                    // ReadException will be caught under the IOException
+                    // - prepare or put throwing this error
+                    // - will move to next url
+                    log.info("transient exception." + te.getMessage());
+
+                } catch (ResourceNotFoundException | ResourceAlreadyExistsException | PreconditionFailedException badURL) {
+                    log.info("removing URL " + u + " from list: " + badURL.getMessage());
```

Comment:
PreconditionFailedException probably means the URL should be discarded. With the code above it should be, but if the PreconditionFailedException was thrown by StorageAdapter.put it could have been a corrupt download that didn't trigger a ReadException (premature EOF maybe)... still, let's stick with discarding the URL for now.

Reply:
ok